### ISSUE TITLE
ShadowCss.shimCssText slow after RC6 update

### ISSUE THREAD

**Issue by Zeetah** [angular/angular#11371]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

https://github.com/angular/angular/blob/38069aba35ac2914bb0a366036271a1388f7c275/modules/%40angular/compiler/src/shadow_css.ts#L151

The above line takes a few hundred milliseconds to run. See the attached JS profile:

<details>
![screen shot 2016-09-06 at 14 18 05](https://cloud.githubusercontent.com/assets/2276092/18271811/cba34238-743c-11e6-8a87-1bb8023b71b8.png)
</details>

If we include images as data urls in the css output, the performance is destroyed. The application startup might take anything from 5 to 30 seconds when having approx. 50 kB css included as component styles.

<details>
![screen shot 2016-09-06 at 13 18 19](https://cloud.githubusercontent.com/assets/2276092/18271843/0b1e1eb0-743d-11e6-8d35-9e3d7ca92c1f.png)
</details>

**Expected/desired behavior**

Shouldn't take that long, RC5 did not have this issue.

**Reproduction of the problem**

Sample webpack configuration snippet

```
module: {
  loaders: [
    {test: /\.ts$/, exclude: [/\.spec\.ts$/, /node_modules/], loader: 'ts'},
    {test: /\.css$/, loader: 'raw'},
    {test: /\.html$/, loader: 'raw'},
    {test: /\.scss$/, loader: 'exports-loader?module.exports.toString()!css!postcss-loader!sass'},
    {test: /\.(png|jpg|gif)$/, exclude: /favicon\//, loader: 'url-loader?limit=8096&name=' + config.imageAssetPath }
  ],
  // ...
}
```

Component

```
@Component({
  styles: [require('./my.component.scss')]
})
class MyComponent {}
```

.scss (output is 70kB when minified)

```
$weather-map: (
  n600: "\e600",
  n500: "\e601",
  n440: "\e602",
  /* ~60 rows */
)
@each $name, $code in $weather-map {
  .weather-#{$name}:before {
    background-image: url('../../../images/weather-icons/#{$name}.png');
  }
}
```

**Please tell us about your environment:**

* **Angular version:** 2.0.0-rc.6
 
* **Browser:** all
 
* **Language:** TypeScript 2.0.2

**Comment by tobias74** [comment]
there is another issue concerning slowness of rc6:

[https://github.com/angular/angular/issues/11334](https://github.com/angular/angular/issues/11334)

**Comment by JohnLGI** [comment]
I was able to work around this problem by commenting out all of my 'styleUrls' declarations in my components and loading all of my css files using the standard <link> tag in my base index.html file.

**Comment by vicb** [comment]
Not sure what the pb is here.

Could you please try to use no encapsulation and see if perfs improve ?

It's RC6 that CSS shimming has been modified in RC6 in order to correctly scope CSS.

Our plan to to move out from the current shimming implementation. @matsko will be working on this - a bench would be interesting.

/cc @tbosch

**Comment by sheniff** [comment]
I can confirm that, when not using encapsulation, performance is restored to < 1s load vs 30+ secs when encapsulating. Any suggested workaround (anything better than having to avoid encapsulation in all my components)?

**Comment by vicb** [comment]
@sheniff how many css lines do you have in your project ?

The overhead is compile time only - AoT will not include this time.

Could you consider having some global styles ? (ie can `.weather-#{$name}` be global ?)

**Comment by sheniff** [comment]
I can consider having some of those styles global, yes. But doesn't it break the idea of angular components to work as web components in terms of isolating their own styles and stuff?

Still, I can wait for a further version where such performance issue is fixed while I develop in rc5.

Thanks!

**Comment by tobias74** [comment]
Changing this:
`        StyleCompiler.prototype._shimIfNeeded = function (style, shim) {
            return shim ? this._shadowCss.shimCssText(style, CONTENT_ATTR, HOST_ATTR) : style;
        };
`
to this:
`        StyleCompiler.prototype._shimIfNeeded = function (style, shim) {
            return style;
        };
`
brings back rc5-performance

**Comment by dcg** [comment]
@tobias74  the issue with this fix for me is that the host selectors don't work anymore.
But i can confirm, that the the regex is using a lot of cpu and time 
![slow](https://cloud.githubusercontent.com/assets/948364/18302780/f5399ef6-74d9-11e6-931e-b6880c0f2743.jpg)

**Comment by matsko** [comment]
So the plan for the shimming is to have it be handled by Angular's CSS parser. This will take a few weeks of work, but the end result will be much better overall since we are able to do a better job since we're not relying on regex-based string replacement.

Even though the current regex solution may be slow at times, keep in mind that you can run the website through the AoT (ngc) compilation. The end result will have the compiled (shimmed) CSS content ready to be used (no compilation needed at runtime).

**Comment by dcg** [comment]
The problem for me is more in the development-time, to wait 30sec for a page reload is a  pitty ;)

atm  i just removed  the  source-map extraction: brings back performance and i don't need the feature atm.

```
./node_modules/@angular/compiler/bundles/compiler.umd.js  Line: 16203

 // var sourceMappingUrl = extractSourceMappingUrl(cssText);  //FIXME
 var sourceMappingUrl = "";
```

**Comment by vicb** [comment]
@tobias74 @dcg thanks for the analysis.

we'll do our best to include a fix in the next release as we understand it is important for dev mode.

**Comment by vicb** [comment]
Could please one of you try the following and report the outcome:

```
// all comments except inline source mapping ("/* #sourceMappingURL= ... */")
const _sourceMappingUrlRe = /\/\*\s*#\s*sourceMappingURL=[\s\S]+?\*\//;

function extractSourceMappingUrl(input: string): string {
  const matcher = input.match(_sourceMappingUrlRe);
  return matcher ? matcher[0] : '';
}
```

should remove backtracking and restore perfs.

_(Same thing that cause [a stack overflow outage](http://stackstatus.net/post/147710624694/outage-postmortem-july-20-2016) some time ago)_

**Comment by dcg** [comment]
@vicb 
Time spend before your fix: 10.48 sec
Time spend after yoru fix: 2.85 sec

looks good to me

**Comment by vicb** [comment]
@dcg thanks for the feedback - the fix has been merged
